hollow/src/main/java/com/netflix/hollow/api/consumer/metrics/AbstractRefreshMetricsListener.java

```diff
@@ -0,0 +1,178 @@
+    private Instant refreshStart;
+
+    private ConsumerRefreshStatus refreshStatus;                    // success or failure
+    private ConsumerRefreshTransitionType refreshTransitionType;    // snapshot or delta transition
+    private ConsumerRefreshLoadType refreshLoadType;                // "init" indicates first successful load on consumer, "subsequent" indicates subsequent load during refresh
+
+    private long refreshLoadCount;
+
+    // Encapsulate the refresh metrics so that more metrics/properties can be added in future wihtout breaking existing clients.
+    public static class ConsumerRefreshMetrics {
+        private long refreshDurationMillis;
+        private ConsumerRefreshStatus refreshStatus;
+        private ConsumerRefreshTransitionType refreshTransitionType;
+        private ConsumerRefreshLoadType refreshLoadType;
+
+        public ConsumerRefreshMetrics(long refreshDurationMillis, ConsumerRefreshStatus refreshStatus, ConsumerRefreshTransitionType refreshTransitionType, ConsumerRefreshLoadType refreshLoadType) {
+            this.refreshDurationMillis = refreshDurationMillis;
+            this.refreshStatus = Objects.requireNonNull(refreshStatus);
+            this.refreshTransitionType = refreshTransitionType;   // Transition type can be unknown for some failed snapshot or delta transition
+            this.refreshLoadType = Objects.requireNonNull(refreshLoadType);
+        }
+
+        public long getRefreshDurationMillis() {
+            return refreshDurationMillis;
+        }
+        public ConsumerRefreshStatus getRefreshStatus() {
+            return refreshStatus;
+        }
+        public ConsumerRefreshTransitionType getRefreshTransitionType() {
+            return refreshTransitionType;
+        }
+        public ConsumerRefreshLoadType getRefreshLoadType() {
+            return refreshLoadType;
+        }
+    }
+
+    public enum ConsumerRefreshTransitionType {
+        Snapshot("snapshot"),       // refresh involved snapshot transition
+        Delta("delta");             // refresh applied a delta transition
+
+        final private String type;
+        ConsumerRefreshTransitionType(String type) {
+            this.type = type;
+        }
+    }
+
+    public enum ConsumerRefreshLoadType {
+        Init("init"),               // consumer's first-time initialized with a snapshot
+        Subsequent("subsequent");   // consumer subsequent updated with either delta or snapshot
+
+        final private String type;
+        ConsumerRefreshLoadType(String type) {
+            this.type = type;
+        }
+    }
+
+    public enum ConsumerRefreshStatus {
+        Success("success"),
+        Failure("failure");
+
+        final private String status;
+        ConsumerRefreshStatus(String status) {
+            this.status = status;
+        }
+    }
+
+
+    public AbstractRefreshMetricsListener() {
+        lastRefreshInstant = Instant.now(); // TODO: Is this an acceptable value for the first transition?
```

Comment:
> Is this an acceptable value for the first transition?

We probably need a sentinel value. Since this is used to report a "time since last refresh" metric, until the initial load we probably shouldn't report that metric at all. For our use, we'd probably want this to be `null` and only when non-null would we start sending the metric.

Looked at another way, this metric should only report zero when it successfully completed an update; viewing a visualization of this metric the troughs should correspond to initial load or subsequent refreshes.

Reply:
ACK